### ISSUE TITLE
Polish Play VoWIFI

### ISSUE THREAD

**Issue by adryb** [kyujin-cho/pixel-volte-patch#17]
Hi, thanks for your app! 

Polish operator - Orange - works well! VoWIFI and VoLTE - great!
Polish operator - Plus - works well! VoWIFI and VoLTE - great!

Polish operator - Play - works well! VoLTE - but no VoWIFi.

I see that there is trying to connect to edge operator server and after some handshake connection dropped logs below.

Maybe some is verified ? Maybe usser agent? I see this can be modified - how to check other User Agent? For example from working samsung?

TCPDUMP:
16:41:47.879021 IP 192.168.253.195.47741 > 89.108.200.112.500: isakmp: parent_sa ikev2_init
16:41:47.919832 IP 89.108.200.112.500 > 192.168.253.195.47741: isakmp: parent_sa ikev2_init[R]
16:41:47.942979 IP 192.168.253.195.43066 > 89.108.200.112.500: isakmp: parent_sa ikev2_init
16:41:47.983686 IP 89.108.200.112.500 > 192.168.253.195.43066: isakmp: parent_sa ikev2_init[R]
16:41:48.016234 IP 192.168.253.195.53081 > 89.108.200.112.4500: NONESP-encap: isakmp: child_sa ikev2_auth
16:41:48.163529 IP 89.108.200.112.4500 > 192.168.253.195.53081: NONESP-encap: isakmp: child_sa ikev2_auth[R]
16:41:48.315983 IP 192.168.253.195.53081 > 89.108.200.112.4500: NONESP-encap: isakmp: child_sa ikev2_auth
16:41:48.406161 IP 89.108.200.112.4500 > 192.168.253.195.53081: NONESP-encap: isakmp: child_sa ikev2_auth[R]
16:41:48.417290 IP 192.168.253.195.53081 > 89.108.200.112.4500: NONESP-encap: isakmp: child_sa ikev2_auth
16:41:48.481634 IP 89.108.200.112.4500 > 192.168.253.195.53081: UDP-encap: ESP(spi=0x59cfb5b6,seq=0x1), length 168
16:41:48.481688 IP 89.108.200.112.4500 > 192.168.253.195.53081: NONESP-encap: isakmp: child_sa ikev2_auth[R]
16:41:49.730284 IP 192.168.253.195.53081 > 89.108.200.112.4500: UDP-encap: ESP(spi=0x051b016c,seq=0x1), length 1304
16:41:49.730285 IP 192.168.253.195.53081 > 89.108.200.112.4500: UDP-encap: ESP(spi=0x051b016c,seq=0x2), length 1128
16:41:49.795468 IP 89.108.200.112.4500 > 192.168.253.195.53081: UDP-encap: ESP(spi=0x59cfb5b6,seq=0x2), length 168
16:41:50.104259 IP 89.108.200.112.4500 > 192.168.253.195.53081: UDP-encap: ESP(spi=0x59cfb5b6,seq=0x3), length 1464
16:41:50.169513 IP 192.168.253.195.53081 > 89.108.200.112.4500: UDP-encap: ESP(spi=0x051b016c,seq=0x3), length 168
16:41:50.193627 IP 89.108.200.112.4500 > 192.168.253.195.53081: UDP-encap: ESP(spi=0x59cfb5b6,seq=0x4), length 936
16:41:50.219036 IP 89.108.200.112.4500 > 192.168.253.195.53081: UDP-encap: ESP(spi=0x59cfb5b6,seq=0x5), length 792
16:41:50.223319 IP 192.168.253.195.53081 > 89.108.200.112.4500: UDP-encap: ESP(spi=0x051b016c,seq=0x4), length 168
16:41:50.246168 IP 192.168.253.195.53081 > 89.108.200.112.4500: UDP-encap: ESP(spi=0x051b016c,seq=0x5), length 168
16:41:50.246168 IP 192.168.253.195.53081 > 89.108.200.112.4500: UDP-encap: ESP(spi=0x051b016c,seq=0x6), length 168
16:41:50.249155 IP 192.168.253.195.53081 > 89.108.200.112.4500: UDP-encap: ESP(spi=0x051b016c,seq=0x7), length 168
16:41:50.265083 IP 192.168.253.195.53081 > 89.108.200.112.4500: UDP-encap: ESP(spi=0x051b016c,seq=0x8), length 168
16:41:50.267608 IP 192.168.253.195.53081 > 89.108.200.112.4500: UDP-encap: ESP(spi=0x051b016c,seq=0x9), length 168
16:41:50.282822 IP 192.168.253.195.53081 > 89.108.200.112.4500: UDP-encap: ESP(spi=0x051b016c,seq=0xa), length 120
16:41:50.295657 IP 89.108.200.112.4500 > 192.168.253.195.53081: UDP-encap: ESP(spi=0x59cfb5b6,seq=0x6), length 168
16:41:50.295713 IP 89.108.200.112.4500 > 192.168.253.195.53081: UDP-encap: ESP(spi=0x59cfb5b6,seq=0x7), length 168
16:41:50.317985 IP 89.108.200.112.4500 > 192.168.253.195.53081: UDP-encap: ESP(spi=0x59cfb5b6,seq=0x8), length 168
16:41:50.318042 IP 89.108.200.112.4500 > 192.168.253.195.53081: UDP-encap: ESP(spi=0x59cfb5b6,seq=0x9), length 168
16:41:50.359287 IP 192.168.253.195.53081 > 89.108.200.112.4500: NONESP-encap: isakmp: child_sa inf2
16:41:50.399064 IP 89.108.200.112.4500 > 192.168.253.195.53081: NONESP-encap: isakmp: child_sa inf2[R]

**Comment by JaCzekanski** [comment]
Can confirm the same behavior on Pixel 6 with Play SIM card - VoLTE works (it actually did without this hack), VoWIFI does not.

**Comment by bige93** [comment]
Same here with Pixel 6 running on T-Mobile Poland network.

Since Android 13 VoLTE works on Pixel 6 out of the box, but not WiFi Calling.

I followed your instructions and run the script, then switched WiFi Calling on in the app.
WiFi Calling setting appeared right away, however I couldn't force it to work even after few reboots.
On top of that, whenever WiFi Calling is enabled in the device settings, phone can't register for VoLTE and calls are being placed over 3G.

Carrier supports WiFi Calling and I have it enabled on my line (I was using it with iPhone 7 previously).

Thanks!

**Comment by adryb** [comment]
I had a chance to check the same - with root and Magisk module - the same with Play - VoLTE works, VoWIFI trying to connect as with your application.

@kyujin-cho can you help me with troubleshoot? I can do this on rooted firmware with root access.

**Comment by kyujin-cho** [comment]
Hi! Thanks for the report. Retrieving IMS User-Agent varies by manufacturer. For example, in Pixel, you need to dial *#*#0702#*#* from default dialer and look up for IMS_USER_AGENT.

**Comment by adryb** [comment]
On my pixel - stock rom - this code not works -> ##0702## its going as ussd.

**Comment by JaCzekanski** [comment]
> On my pixel - stock rom - this code not works -> ##0702## its going as ussd.

@adryb I think it's supposed to be `*#*#0702#*#*`

**Comment by adryb** [comment]
Ok on google its:

#MANUFACTURE#_#MODEL#_Android #AV#_#BUILD#

**Comment by dadi11** [comment]
For me on Pixel 6 and Play sim it's 
`#MANUFACTURE#_#MODEL#_Android #AV#_#BUILD#`

**Comment by adryb** [comment]
I made a lot of tests, on rooted phone too. Tried to capture traffic on phone side - in ipsec tunnel there is sedond tunnel that communicates with operator devces, i see only that fter some handshake(?) operator closes connection.

Its not related to patch.

**Comment by kszaq** [comment]
@adryb Can you post a result of `adb shell dumpsys carrier_config` with Play SIM?

**Comment by adryb** [comment]
Sure. Attached.
[cc_pixel.log](https://github.com/kyujin-cho/pixel-volte-patch/files/10774136/cc_pixel.log)

**Comment by kszaq** [comment]
Thanks. The only difference I can see between Orange PL (working) and Play (not working) is that for Play, "ims" APN is defined in carrier settings. Can you try to add APN "ims" with any name and only APN type set to "ims", other options at default?

**Comment by adryb** [comment]
Yep! That worked!!!!! IM ABLE TO CALL VIA VOWIFI!!!!

**Comment by kszaq** [comment]
It's not a line from `carrier_config`. I looked at carrier settings in `/system/product/etcCarrierSettings/`. You can decode them using `protoc --decode_raw`. For Play, `ims` APN is defined for only some bearers, I assume LTE and UMTS and because of this it's not used for IWLAN. If you redefine `ims` with no bearer, it is used for all network types and makes IMS register over WLAN.

**Comment by NXij** [comment]
`bearer_bitmask: "18"` should be for IWLAN then yeah. Not seeing a way to change it through the Carriersettings api.

**Comment by dadi11** [comment]
> Yep! That worked!!!!! IM ABLE TO CALL VIA VOWIFI!!!!
> 
> Which line this shows you?

How did it work for you? I've added ims APN and IMS is still not registering over WiFi

**Comment by adryb** [comment]
Yes, its working after phone restart too! Everything is wotking well, i have physical sim Play and esim Plus, both works!

**Comment by JaCzekanski** [comment]
> Thanks. The only difference I can see between Orange PL (working) and Play (not working) is that for Play, "ims" APN is defined in carrier settings. Can you try to add APN "ims" with any name and only APN type set to "ims", other options at default?

Thank you, that actually worked!

Here are the settings for APN if anyone having issues:
![Screenshot_20230220-153023](https://user-images.githubusercontent.com/3662990/220134391-0deb9f8c-2e1e-4b23-b7d9-6976793269de.png)

**Comment by exolon** [comment]
Rats, not even this worked for me (I'm in Greece btw)... VolTE is provisioned, there is the option in the SIM settings (and turned on of course) but everything is not registered or unavailable... Mainly, IMS Registration is Not Registered, even with the VPN :( Any other ideas? Pixel 7 Pro here.

**Comment by ciarancoffey** [comment]
The APN method worked for me on GoMo in Ireland.

**Comment by bige93** [comment]
Tried adding ims APN and WiFi calling indeed works in Pixel 6a on Play.
The only issue is device still thinks it's on VoWiFi for like a minute after WiFi is disconnected and requires airplane mode toggle to get it back on VoLTE. 
Pixel IMS app is courtesy of @kszaq from this[this](https://github.com/kyujin-cho/pixel-volte-patch/discussions/63#discussioncomment-5241545) post.

**Comment by lepricon** [comment]
For Pixel6 TMobile VoLTE works with both stock and PixelIMS. But VoWiFi doesn't, even with APN set to IMS. Additionally no mobile internet with changed APN. How to debug this for T-Mobile case?

**Comment by danparidae** [comment]
> For Pixel6 TMobile VoLTE works with both stock and PixelIMS v1.2.3. But VoWiFi doesn't (unregisters in a moment), even with APN set to IMS. Additionally no mobile internet with changed APN. How to debug this for T-Mobile case?

You've messed up adding this APN. Try to remove it and add again exactly as it was supposed to look, like on screenshots above

**Comment by lepricon** [comment]
> > For Pixel6 TMobile VoLTE works with both stock and PixelIMS v1.2.3. But VoWiFi doesn't (unregisters in a moment), even with APN set to IMS. Additionally no mobile internet with changed APN. How to debug this for T-Mobile case?
> 
> You've messed up adding this APN. Try to remove it and add again exactly as it was supposed to look, like on screenshots above

Right! Only three should be changed (Name, APN, APN Type), but lots are displayed. Works with T-Mobile also. Thanks

**Comment by hvandenbroeck** [comment]
Creating the APN worked for Orange Belgium + pixel 6 as well.
Thank you!

**Comment by mano922** [comment]
Hello, 
I have a problem with VoWiFi in Orange Poland in Pixel 7 Pro.
VoLTE works fine.
VoWiFi is working with my SIM card on different phone (Xiaomi)

On Pixel 7 Pro I have applied patch, VoWiFi enabled in options but I am not able to make calls with VoWiFI

Added "ims" APN as shown above but it does not help (only three options changed as mentioned above). 
Now I have 3 APNs: 2 defaults: "Orange Internet", "Orange MMS" and new "ims".

Do you have maybe any idea what can be still missing?

**Comment by kszaq** [comment]
> Hello, 
> I have a problem with VoWiFi in Orange Poland in Pixel 7 Pro.
> VoLTE works fine.
> VoWiFi is working with my SIM card on different phone (Xiaomi)
> 
> On Pixel 7 Pro I have applied patch, VoWiFi enabled in options but I am not able to make calls with VoWiFI
> 
> Added "ims" APN as shown above but it does not help (only three options changed as mentioned above). 
> Now I have 3 APNs: 2 defaults: "Orange Internet", "Orange MMS" and new "ims".
> 
> Do you have maybe any idea what can be still missing?

For Orange PL you don't need to add IMS APN - you can reset APN settings to default to revert this. Check in Settings what is your WiFi Calling preference, it may default to calling over mobile unless it's not available.

**Comment by mano922** [comment]
> > Hello,
> > I have a problem with VoWiFi in Orange Poland in Pixel 7 Pro.
> > VoLTE works fine.
> > VoWiFi is working with my SIM card on different phone (Xiaomi)
> > On Pixel 7 Pro I have applied patch, VoWiFi enabled in options but I am not able to make calls with VoWiFI
> > Added "ims" APN as shown above but it does not help (only three options changed as mentioned above).
> > Now I have 3 APNs: 2 defaults: "Orange Internet", "Orange MMS" and new "ims".
> > Do you have maybe any idea what can be still missing?
> 
> For Orange PL you don't need to add IMS APN - you can reset APN settings to default to revert this. Check in Settings what is your WiFi Calling preference, it may default to calling over mobile unless it's not available.

Thanks a lot for your comment and help.
Unfortunately it does not solve the problem.
I removed IMS APN by resetting APN settings and restarted the phone.

I have tested it in airplane mode as well as with the "WiFi only" setting for WiFi calling. In both cases it does not work and when trying to make a call the message: "Wireless networks are disabled" is displayed and it is not possible to make a call.

**Comment by kszaq** [comment]
@mano922 I compared your carrier config with mine and the only difference is that you have `qns.block_ipv6_only_wifi_bool = true` in `mConfigFromDefaultApp`, which seems to be specific to Pixel 7. Today evening, I'll try to build an APK for you with an option to disable this.

**Comment by mano922** [comment]
> @mano922 I compared your carrier config with mine and the only difference is that you have `qns.block_ipv6_only_wifi_bool = true` in `mConfigFromDefaultApp`, which seems to be specific to Pixel 7. Today evening, I'll try to build an APK for you with an option to disable this.

Thanks a lot.

**Comment by mano922** [comment]
> @mano922 You can try this APK: [app-debug-631c8db.zip](https://github.com/kyujin-cho/pixel-volte-patch/files/11161611/app-debug-631c8db.zip) You need to unpack it and remove Kyujin-Cho's APK before installation because of signature mismatch. Please note that it is from my custom branch with some additional toggles. Try to disable "Block IPv6-only WiFi".

Thanks a lot for providing your APK.
Unfortunately this updated APK does not solve the problem and the behavior is exactly the same as previously (WiFi calling does not work and VoLTE works fine).

When WiFi calling is activated in settings (options "Use Wifi" or "Wifi only") then IMS status is unregistered and VoLTE unavailable. When WiFi calling is deactivated in settings then IMS status is registered and VoLTE available. 

Please find attached new carrier config. It looks that qns.block_ipv6_only_wifi_bool is set to false.
[logCarrierUpd.txt](https://github.com/kyujin-cho/pixel-volte-patch/files/11162971/logCarrierUpd.txt)

**Comment by kszaq** [comment]
Unfortunately, at this point I am out of ideas what could possibly help. The same carrier config is working for me on Pixel 6 with Orange PL SIM, with latest QPR Beta.

I also remember that at some point in the past @kyujin-cho mentioned that you may need to use "Reset Wi-Fi, mobile & Bluetooth". If you can forget these settings, it may be worth a try.

**Comment by mano922** [comment]
> Unfortunately, at this point I am out of ideas what could possibly help. The same carrier config is working for me on Pixel 6 with Orange PL SIM, with latest QPR Beta.
> 
> I also remember that at some point in the past @kyujin-cho mentioned that you may need to use "Reset Wi-Fi, mobile & Bluetooth". If you can forget these settings, it may be worth a try.

@kszaq thanks a lot for your help and hints
Unfortunately after doing the "Reset Wi-Fi, mobile & Bluetooth" the behavior is still the same as previously and VoWiFi does not work.

**Comment by kszaq** [comment]
@mano922 I've installed the latest QPR Beta on my Pixel 6 and now VoLTE is available by default. But for VoWiFi, I needed to add IMS APN. One thing to note: it's best to set APN protocols to IPv4/IPv6 because for Orange PL you need IPv6. IPv4 did not work for me.

**Comment by mano922** [comment]
@kszaq Thanks a lot for your hints.

I have also installed yesterday latest QPR Beta on my Pixel 7 Pro and VoLTE is also available by default for me.

Unfortunately VoWiFI does not work (the same behavior as previously). Tested with different options: airplane mode, WiFi only setting, Orange PL with APN ims with protocol IPv4/IPv6, "Block IPv6-only WiFi" disabled.

**Comment by mr-tical** [comment]
I do have the same situation with PL Plus - VoLTE works, VoWiFi stopped working. Currently on latest qpr beta. VoLTE appeared itself after updating to qpr beta, before was not present and had to use this patch 
On stable March update VoWiFi worked but not VoLTE.
But on January and February update VoWiFi and VoLTE worked with the patch.
Now on qpr beta I see the carrier settings updated to pl-plus 2023-03-24 also.
Tried adding also ims APN, but when added, VoLTE speed working immediately.
Had to remove it and reboot phone.
Using Pixel 7.

**Comment by mr-tical** [comment]
Patch is working on Android 14 beta 1
VoWiFi and VoLTE 
PL Plus

**Comment by fundusze** [comment]
Hi, I can't configure out Pixel IMS service on Pixel 7 (Virgin mobile PL - now a part of PLAY carrier). I tried with/without ims APN but still without success.![Screenshot_20230906-223653.png](https://github.com/kyujin-cho/pixel-volte-patch/assets/22707317/70448e32-646f-4327-a2bf-7feb718134f9)
![Screenshot_20230906-224924.png](https://github.com/kyujin-cho/pixel-volte-patch/assets/22707317/ccccf97d-3724-46b5-8254-2d5d499cc98b)

![Screenshot_20230906-224915.png](https://github.![Screenshot_20230906-224906.png](https://github.com/kyujin-cho/pixel-volte-patch/assets/22707317/14861ba6-6390-4840-b11e-b4aae3b2b973)

848302cc-fea5-4915-8596-f6d7fcdb1688)

![Uploading Screenshot_20230906-224906.png …]()

**Comment by fillwithjoy1** [comment]
> Same here with Pixel 6 running on T-Mobile Poland network.
> 
> Since Android 13 VoLTE works on Pixel 6 out of the box, but not WiFi Calling.
> 
> I followed your instructions and run the script, then switched WiFi Calling on in the app. WiFi Calling setting appeared right away, however I couldn't force it to work even after few reboots. On top of that, whenever WiFi Calling is enabled in the device settings, phone can't register for VoLTE and calls are being placed over 3G.
> 
> Carrier supports WiFi Calling and I have it enabled on my line (I was using it with iPhone 7 previously).
> 
> Thanks!

funny thing is that wifi calling on the iphone is a different platform to support than to wifi calling on android

**Comment by fillwithjoy1** [comment]
also is this issue relevant now?

**Comment by fillwithjoy1** [comment]
@kyujin-cho requesting to close the issue

**Comment by szmarczak** [comment]
I'm using nju mobile and WiFi calling doesn't work even after adding the ims apn (nothing phone 2).

Edit: I tried a carrier certified device (Samsung a53) and it still does not work. So my case is not relevant here and I need to contact my carrier's support.

**Comment by fillwithjoy1** [comment]
> I'm using nju mobile and WiFi calling doesn't work even after adding the ims apn (nothing phone 2).
> 
> Edit: I tried a carrier certified device (Samsung a53) and it still does not work. So my case is not relevant here and I need to contact my carrier's support.

Well, how interesting that it's the carrier's issue...

**Comment by samuelgarabini** [comment]
@kszaq Can you help me with my problem? https://github.com/kyujin-cho/pixel-volte-patch/issues/323

**Comment by kszaq** [comment]
First make sure that it is working with your SIM on another phone. Sometimes operators need VoWiFi to be activated by accepting terms & conditions.

**Comment by samuelgarabini** [comment]
> First make sure that it is working with your SIM on another phone. Sometimes operators need VoWiFi to be activated by accepting terms & conditions.

Yes, it it is working. When I try in a different phone, both VoWifi and VoLTE works, but mine only VoLTE. I tried everything today. I appreacite if you could help me

**Comment by kszaq** [comment]
The only thing I can recommend is that you delete "IMS" APN and re-add it, making sure that "bearer" is "not specified" only. If this doesn't work, it is possible that your network operator does not allow your phone to use VoWiFi.

**Comment by samuelgarabini** [comment]
> The only thing I can recommend is that you delete "IMS" APN and re-add it, making sure that "bearer" is "not specified" only. If this doesn't work, it is possible that your network operator does not allow your phone to use VoWiFi.

What do I need to mark besides "not specified"? Everything?
